cheatsheets/Nodejs_security_cheat_sheet.md

```diff
@@ -0,0 +1,371 @@
+
+```JavaScript
+exports.sanitizeUser = function(user) {
+  return {
+    id: user.id,
+    username: user.username,
+    fullName: user.fullName
+  };
+};
+```
+
+## Set request size limits
+
+Buffering and parsing of request bodies can be cumbersome for the server. If there is no limit on the size of requests, attackers can send request with large request bodies so that they can exhaust server memory or fill disk space. However, fixing a request size limit for all requests may not be the correct behavior, since some requests like those for uploading a file to the server have more content to carry on the request body. Also, input with a JSON type is more dangerous than a multipart input, since parsing JSON is a blocking operation. Therefore, you should set request size limits for different content types. You can accomplish this very easily with express middlewares as follows:
+
+```JavaScript
+app.use(express.urlencoded({ limit: “1kb” }));
+app.use(express.json({ limit: “1kb” }));
+app.use(express.multipart({ limit:”10mb” }));
+app.use(express.limit(“5kb”)); // this will be valid for every other content type
+```
+
+## Use strict mode
+
+JavaScript has some unsafe features that lies within the language itself. In order to remove these features, ES5 included a strict mode for developers. With this mode, errors that were silent previously are thrown. It also restricted optimizations for JavaScript engines. With strict mode, previously accepted bad syntax causes real errors. use "use strict". Because of these improvements, you should always use strict mode in your application. In order to enable strict mode, you just need to write _‘use strict’;_ on top of your code.
+
+## Use object property descriptors
+
+Object properties include 3 hidden attributes: writable (if false, property value cannot be changed), enumerable (if false, property cannot be used in for loops) and configurable (if false, property cannot be deleted). When defining an object property through assignment, these three hidden attributes are set to true by default. These properties can be set as follows:
+
+```JavaScript
+var o = {};
+Object.defineProperty(o, “a”, {
+    writable: true,
+    enumerable: true,
+    configurable: true,
+    value: “A”
+});
+```
+
+Apart from these, there are some special functions for object attributes. Object.preventExtensions() prevents new properties from being added to the object.
+
+## Handle errors in asynchronous calls
+
+Errors that occur within asynchronous callbacks can be easily forgettable. Therefore, as a general principle first argument to the asynchronous calls should be an Error object. Also, express routes handle errors itself, but it should be always remembered that errors occurred in asynchronous calls made within express routes are not handled, unless an Error object is sent as a first argument.
```

Comment:
`are easy to miss` rather than `are easily forgettable`

Reply:
modified accordingly